src/session-pool.js

```diff
@@ -75,874 +129,662 @@ const READWRITE = 'readwrite';
-  this.pingHandle_ = null;
-
-  this.traces_ = new Map();
-
-  EventEmitter.call(this);
-}
-
-util.inherits(SessionPool, EventEmitter);
-
-/**
- * Formats stack trace objects into Node-like stack trace.
- *
- * @private
- *
- * @param {object[]} trace The trace object.
- * @return {string}
- */
-SessionPool.formatTrace_ = function(trace) {
-  const formatted = trace.slice(2).map(function(t) {
-    return `    at ${t.getFunctionName() ||
-      t.getMethodName()} (${t.getFileName()}:${t.getLineNumber()}:${t.getColumnNumber()})`;
-  });
-
-  return `Session leak detected!\n${formatted.join('\n')}`;
-};
-
-/**
- * Returns the number of available sessions.
- *
- * @return {number}
- */
-SessionPool.prototype.available = function() {
-  return this.reads_.length + this.writes_.length;
-};
-
-/**
- * Returns the number of borrowed sessions.
- *
- * @return {number}
- */
-SessionPool.prototype.borrowed = function() {
-  return this.borrowed_.length;
-};
+  get borrowed() {
+    return this._inventory.borrowed.size;
+  }
 
-/**
- * Closes and the pool.
- *
- * @emits SessionPool#close
- * @returns {Promise}
- */
-SessionPool.prototype.close = function() {
-  const self = this;
-  const sessions = this.reads_.concat(this.writes_, this.borrowed_);
+  /**
+   * Flag to determine if Pool is full.
+   * @type {boolean}
+   */
+  get isFull() {
+    return this.size >= this.options.max;
+  }
 
-  this.isOpen = false;
+  /**
+   * Total number of read sessions.
+   * @type {number}
+   */
+  get reads() {
+    const {readonly, borrowed} = this._inventory;
+    const available = readonly.length;
+    const used = Array.from(borrowed).filter(
+      session => session.type === READONLY
+    ).length;
+
+    return available + used;
+  }
 
-  this.reads_ = [];
-  this.writes_ = [];
-  this.borrowed_ = [];
+  /**
+   * Total size of pool.
+   * @type {number}
+   */
+  get size() {
+    return this.available + this.borrowed;
+  }
 
-  this.stopHouseKeeping_();
+  /**
+   * Total number of write sessions.
+   * @type {number}
+   */
+  get writes() {
+    const {readwrite, borrowed} = this._inventory;
+    const available = readwrite.length;
+    const used = Array.from(borrowed).filter(
+      session => session.type === READWRITE
+    ).length;
+
+    return available + used;
+  }
 
-  this.emit('empty');
-  this.emit('close');
+  /**
+   * @constructor
+   * @param {Database} database The DB instance.
+   * @param {SessionPoolOptions} [options] Configuration options.
+   */
+  constructor(database, options) {
+    super();
 
-  return Promise.all(
-    sessions.map(function(session) {
-      return self.destroySession_(session);
-    })
-  );
-};
+    this.isOpen = false;
+    this.database = database;
+    this.options = Object.assign({}, DEFAULTS, options);
 
-/**
- * Fills the pool with the minimum number of sessions.
- *
- * @return {Promise}
- */
-SessionPool.prototype.fill = function() {
-  const requests = [];
-  let i;
+    if (this.options.writes > 1) {
```

Comment:
should probably check this is `>= 0` too

Reply:
Should the other options have sanity checks as well?  Recently ran into a really weird error from passing a mis-named option, this could help prevent that.

Reply:
What kind of error? And which option?

Reply:
As far as options go, `keepAlive` is the only other one that might benefit from a sanity check, because if the value is greater than 60 minutes, the sessions will always die.

Reply:
Was getting errors relating to invalid arguments being passed to what looks like the grpc request to create a spanner session.  I had been passing an options object with unexpected keys.  I'll submit an issue with this if I can reproduce it.